core/src/ast/expr.rs

```diff
@@ -61,3 +61,294 @@ pub enum Expr {
         else_result: Option<Box<Expr>>,
     },
 }
+
+impl ToSql for Expr {
+    fn to_sql(&self) -> String {
+        match self {
+            Expr::Identifier(s) => s.to_string(),
+            Expr::BinaryOp { left, op, right } => {
+                format!("{} {} {}", left.to_sql(), op.to_sql(), right.to_sql())
+            }
+            Expr::CompoundIdentifier(idents) => idents.join("."),
+            Expr::IsNull(s) => format!("{} IS NULL", s.to_sql()),
+            Expr::IsNotNull(s) => format!("{} IS NOT NULL", s.to_sql()),
+            Expr::InList {
+                expr,
+                list,
+                negated,
+            } => {
+                let expr = expr.to_sql();
+                let list = list
+                    .iter()
+                    .map(ToSql::to_sql)
+                    .collect::<Vec<_>>()
+                    .join(", ");
+
+                match negated {
+                    true => format!("{expr} NOT IN ({list})"),
+                    false => format!("{expr} IN ({list})"),
+                }
+            }
+            Expr::Between {
+                expr,
+                negated,
+                low,
+                high,
+            } => {
+                let expr = expr.to_sql();
+                let low = low.to_sql();
+                let high = high.to_sql();
+
+                match negated {
+                    true => format!("{} NOT BETWEEN {} AND {}", expr, low, high),
+                    false => format!("{} BETWEEN {} AND {}", expr, low, high),
+                }
+            }
+            Expr::UnaryOp { op, expr } => format!("{}{}", op.to_sql(), &*expr.to_sql()),
+            Expr::Cast { expr, data_type } => {
+                format!("CAST({} AS {})", &*expr.to_sql(), data_type)
+            }
+            Expr::Extract { field, expr } => {
+                format!("EXTRACT({} FROM \"{}\")", field, &*expr.to_sql())
+            }
+            Expr::Nested(expr) => format!("todo:Nested({})", &*expr.to_sql()),
+            Expr::Literal(s) => s.to_sql(),
+            Expr::TypedString { data_type, value } => format!("{data_type}(\"{value}\")"),
+            Expr::Case {
+                operand,
+                when_then,
+                else_result,
+            } => {
+                let mut str = match operand {
+                    Some(s) => format!("CASE {}", s.to_sql()),
+                    None => "CASE ".to_string(),
+                };
+                for (_when, _then) in when_then {
+                    str += format!("\nWHEN {} THEN {}", _when.to_sql(), _then.to_sql()).as_str();
+                }
+
+                match else_result {
+                    Some(s) => str += format!("\nELSE {}", s.to_sql()).as_str(),
+                    None => str += "", // no operation?
+                };
+                str + "\nEND"
```

Comment:
It would be good not to use `mut` here
In Rust, it is recommended to use variable shadowing because it can explicitly show the lifetime of each variable.

e.g.
```rust
let operand = match operand {
    Some(operand) => format!("CASE {}", operand.to_sql()),
    None => "CASE".to_owned(),
};  
    
let when_then = when_then
    .iter()
    .map(|(when, then)| format!("WHEN {} THEN {}", when.to_sql(), then.to_sql()))
    .collect::<Vec<_>>()
    .join("\n");

let else_result = match else_result {
    Some(else_result) => format!("ELSE {}", else_result.to_sql()),
    None => String::new(),
};  

[operand, when_then, else_result, "END".to_owned()].join("\n")
```

Reply:
Thanks for the code rewrite..  It helps a lot.